### ISSUE TITLE
Bug: [no-shadow] ignoreFunctionTypeParameterNameValueShadow = true false positive on function interface dec

### ISSUE THREAD

**Issue by kavsingh** [typescript-eslint/typescript-eslint#6098]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=KYDwDg9gTgLgBAYwgOwM7xHAvHADAbgChRJZEV04BPbPIwgegbmQjmCimgBo4B3aAGtUcAIYiSwBDGAATYuGjwYVMMDgBJWcGTwcAHgAqAPgAUIAFxxDASmzHr9JizYcuUXgKjCxE8FJl5EiU4AEtdDgAzUQR1AHkAIwArOABvQjgxWVkrcytkAFcAWwSOXip84tKoOywHQpKOIgBfQgVSeCQ0eFQIIuAAMQLkBCt0zOc3aDgAchAZsJFRABsoYFFZGm0EZdE12TDkOBgAC3UCsDUoOFQkNQy4Sc5p1nhJaTkHozNLaxsrQxECbMKbXOYLUJLVbrTZwba7faHY5nOAXK43O7AB5PdyzKgQqFrDZbKQIuRI07nS4cDEQe7A9jPa6vdj+D7yTLfPLWcqVRo1AFwAA+LCqTUIzVoplEvASAH4+dVag4EnA5XKxPRgmRwjIoNFYpohiM0tiQUzZvNFmJocS4aS9uTwsiqejbnSsQzQS43mzAl8TNzbACgY9zbjwdaVkTYfDHQdnZTUdTru76WHGRH8VHbbGHYjEyi0TS056M96We9-ZzA79DLzRfz-tZhY3qi02trlKp1IZjQhaONyxbI5CbTGSTt4xSiynaemcS8IL61OyAz8Ac3AWbM9NR4SYZOyQmjkni6nMTvvTNs2Po4f7VOC6fZ27L16LZW-Z8axuedRFQ4LdWwadsJTaIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uQAsBDACYcA7ujABtcNhyJo0DtEgAaKdKzTskSgHNO0RABUSiAGp94sRAGV+Q0RgBm55IhUbNOvYgBisJmXxKDiYjUgAFPmg+AFtEfFkAORjTc0sbQRExfDhEVWwAXzyAXSlCwqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
export const x = 0;
export const y = 0;

// no error, works as expected
export type Ident = <T>(x: T) => T;

// no error, works as expected
export interface Obj {
  add: (x: number, y: number) => number;
}

export const someFunc: {
  // error 'x' is already declared in the upper scope
  // error not expected
  <T>(x: T): T;
  // error 'x' is already declared in the upper scope
  // error 'y' is already declared in the upper scope
  // error not expected
  <T>(x: T, y: number): T | number;
} = (a, b?: number) => b ?? a;

export interface IFunc {
  // error 'x' is already declared in the upper scope
  // error not expected
  <T>(x: T): T;
  // error 'x' is already declared in the upper scope
  // error 'y' is already declared in the upper scope
  // error not expected
  <T>(x: T, y: number): T | number;
}

export type TFunc = {
  // error 'x' is already declared in the upper scope
  // error not expected
  <T>(x: T): T;
  // error 'x' is already declared in the upper scope
  // error 'y' is already declared in the upper scope
  // error not expected
  <T>(x: T, y: number): T | number;
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-shadow": [
      "error",
      {
        "ignoreTypeValueShadow": false,
        "ignoreFunctionTypeParameterNameValueShadow": true
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

with `ignoreFunctionTypeParameterNameValueShadow` set to true,
expected no errors when shadowing variables in function interface parameters

### Actual Result

with `ignoreFunctionTypeParameterNameValueShadow` set to true,
received errors when shadowing variables in function interface parameters

### Additional Info

was working as expected in v5.42.1, errors as of v5.43.0

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.44.0` |
| `@typescript-eslint/parser` | `5.44.0` |
| `TypeScript` | `4.8.4` |
| `ESLint` | `8.15.0` |
| `node` | `web` |

**Comment by JoshuaKGoldberg** [comment]
`ignoreFunctionTypeParameterNameValueShadow` is specifically for _type_ parameters, not regular (_value_) parameters. In this example, `T` is a _type_ parameter and `value` is a _value_ parameter:

```ts
declare function example<T>(value: T): void;
```

Consider this simplified version of the playground, where `typeof x` refers to the `x: T` parameter, not the `const x` variable.

```ts
declare const x: number;

declare const someFunc: {
  // error 'x' is already declared in the upper scope
  <T>(x: T): typeof x;
};
```

The rule is working as intended here. Thanks for filing!

**Comment by kavsingh** [comment]
thanks for the insight!

**Comment by EvanCahill** [comment]
I think this issue was incorrectly closed and should be re-opened.

The following code errors when `ignorefunctiontypeparameternamevalueshadow` is `false` but is fine when it is `true`.
```ts
const arg = '';
interface Test {
  // reports error on arg when ignorefunctiontypeparameternamevalueshadow is false
  p: (arg: string) => void;
}
```

This clearly shows that `ignorefunctiontypeparameternamevalueshadow` is not for type parameters but for the value parameters. The [documentation](https://typescript-eslint.io/rules/no-shadow/#ignorefunctiontypeparameternamevalueshadow) for `ignoreFunctionTypeParameterNameValueShadow` also makes no mention of type parameters.

Either way, the current behavior is a regression from v4.42.1 to v4.43.0 which appears to have been caused by #5902. The below code works fine in the preview playground of the commit just before the change, but errors in the PR's playground. 

```ts
const arg = '';
interface Test {
  (arg: string): void; // error
  p1(arg: string): void; // error
  p2: (arg: string) => void; // no error
}
```

**Comment by JoshuaKGoldberg** [comment]
Hmm, that looks like a bug for me. Thanks for mentioning 😄. Filed #6125

**Comment by JoshuaKGoldberg** [comment]
Also, you're right - thanks. When `ignoreFunctionTypeParameterNameValueShadow` is true, we should report none of these as complaints.

**Comment by JoshuaKGoldberg** [comment]
Ok - I took a deep dive and realized I'd done a silly mistake and inverted my understanding of the option. Sorry for the back-and-forth folks - you were right @kavsingh & @EvanCahill. My comment https://github.com/typescript-eslint/typescript-eslint/issues/6098#issuecomment-1328113338 is incorrect and I'll hide it so folks don't read it and think it's true.

I'll also send a PR for this one, so another maintainer has to double-check my work. Just in case I'm wrong again!
